lib/amo-client.js

```diff
@@ -37,25 +49,98 @@ function Client(conf) {
  * @return {Promise}
  */
 Client.prototype.sign = function(conf) {
+  var self = this;
   conf = conf || {};
 
-  var apiUrl = '/addons/' + encodeURIComponent(conf.guid);
-  apiUrl += '/versions/' + encodeURIComponent(conf.version) + '/';
+  var addonUrl = '/addons/' + encodeURIComponent(conf.guid);
+  addonUrl += '/versions/' + encodeURIComponent(conf.version) + '/';
 
   return this.put({
-    url: apiUrl,
+    url: addonUrl,
     formData: {
       upload: this._fs.createReadStream(conf.xpiPath),
     },
+  }, {
+    throwOnBadResponse: false,
   }).then(function(responseResult) {
     var httpResponse = responseResult[0];
     var response = responseResult[1];
 
-    console.log('Your add-on file is being processed');
-    console.log('Job ID:', response.id)
-    console.log('Not yet implemented: polling until completion, downloading the signed XPI');
+    var handledStatuses = [200, 409];
+    if (handledStatuses.indexOf(httpResponse.statusCode) === -1) {
+      throw new Error(
+        'Received bad response from the server while signing; ' +
+        'status: ' + httpResponse.statusCode + '; ' +
+        'response: ' +
+        (typeof response === 'string' ? response.substring(0, 100) + '...' : ''));
+    }
+
+    // TODO: maybe return early for 409s (already validated).
+    // Right now they are handled like 200s which seems wrong.
+
+    return self.waitForSignedAddon(addonUrl);
+  });
+};
+
+
+/**
+ * Poll a status URL, waiting for the queued add-on to be signed.
+ *
+ * @param {String} URL to GET for add-on status.
+ * @return {Promise}
+ */
+Client.prototype.waitForSignedAddon = function(statusUrl, opt) {
+  opt = _.assign({
+    clearTimeout: clearTimeout,
+    setAbortTimeout: setTimeout,
+    setStatusCheckTimeout: setTimeout,
+    abortAfter: 45000,
+  }, opt);
+  var self = this;
+
+  return when.promise(function(resolve, reject) {
+    self._validateProgress.animate();
+    var statusCheckTimeout;
+    var nextStatusCheck;
+
+    function checkSignedStatus() {
+      self.get({url: statusUrl}).then(function(result) {
+        var data = result[1];
+        if (data.valid && data.download_url) {
+
+          self._validateProgress.finish();
+          logger.log('SUCCESS');
+          // TODO: actually download the signed file from the URL.
+          logger.log('Your add-on has been signed:', data.download_url);
```

Comment:
`download_url` is actually an array right now. Originally I thought it would be a single URL but there can be multiple files (one for each platform) so they'd each get their own download URL.

An array might not handle multiple files well though since there is some metadata that the user probably wants (what platform it's for at least). So then we could have an array of files or a hash of files keyed off of platform. I'm not sure what is best but `download_url` definitely won't work forever. The API only supports the "ALL" platform right now so maybe I should go back to just `download_url` until we add support.